transaction/std/src/transaction_builder.rs

```diff
@@ -28,16 +29,21 @@ pub struct TransactionBuilder {
     outputs_and_shared_secrets: Vec<(TxOut, RistrettoPublic)>,
     tombstone_block: u64,
     pub fee: u64,
+    fog_pubkeys: FullyValidatedFogPubkeys,
 }
 
 impl TransactionBuilder {
     /// Initializes a new TransactionBuilder.
-    pub fn new() -> Self {
+    ///
+    /// # Arguments
+    /// * `fog_pubkeys` - Set of prefetched fog public keys to choose from
+    pub fn new(fog_pubkeys: FullyValidatedFogPubkeys) -> Self {
         TransactionBuilder {
             input_credentials: Vec::new(),
             outputs_and_shared_secrets: Vec::new(),
-            tombstone_block: u64::max_value(),
+            tombstone_block: fog_pubkeys.smallest_pubkey_expiry_value(),
```

Comment:
In the event that `FullyValidatedFogPubkeys` contains no fog pubkeys and the user doesn't explicity call `TransactionBuilder::set_tombstone_block`, then the built `Tx` will have a `tombstone_block` that will cause it to be rejected by consensus for being too far in the future, meaning that in this situation, `TransactionBuilder::set_tombstone_block` _must_ be called in order to build a valid Tx (which was the case before this PR also).

Because of this, I'm not convinced there's utility in setting the default `tombstone_block` to anything other than `u64::max_value()`. My reasoning is, why default to a value that might or might not be a valid `tombstone_block` in certain circumstances and is definitely not a valid `tombstone_block` in other circumstances? Seems like it would be less confusing to just always set it to a non-valid tombstone block and then have the caller always call `TransactionBuilder::set_tombstone_block`.

Reply:
interesting, i didn't know consensus will reject things for being too far in the future

does consensus just test it against `MAX_TOMBSTONE_BLOCKS` ?

yeah, it seems fine to require that the user set a tombstone value, and we enforce the limit at the time that they try to set it

Reply:
when does the user want to set the tombstone block though? Is there a time that you don't want it to be the smallest of pubkey expiry values that is used? That might be a sane default

Reply:
The smallest of pubkey expiry values won't always be available or be a valid default.

E.g.:
* When all of the recipients are non-fog users.
* When the fog pubkey expiry returned from the fog report server is too far in the future for it to be used as a valid tombstone block.

Reply:
> interesting, i didn't know consensus will reject things for being too far in the future
> 
> does consensus just test it against `MAX_TOMBSTONE_BLOCKS` ?

Exactly. Based on the current ledger block count.

https://github.com/mobilecoinfoundation/mobilecoin/blob/9cf0f47ac841aba6b8e343695f2900a566c70ccf/transaction/core/src/validation/validate.rs#L353

https://github.com/mobilecoinfoundation/mobilecoin/blob/5472a0377a03faddb17c650fb5fd944832ff4cc8/consensus/enclave/impl/src/lib.rs#L408